src/bsoncxx/builder/list/array.hpp

```diff
@@ -0,0 +1,220 @@
+// Copyright 2020 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <bsoncxx/config/prelude.hpp>
+
+#include <bsoncxx/builder/core.hpp>
+#include <bsoncxx/types/bson_value/value.hpp>
+
+namespace bsoncxx {
+BSONCXX_INLINE_NAMESPACE_BEGIN
+namespace builder {
+namespace list {
+using namespace bsoncxx::types;
+
+///
+/// A JSON-like builder for creating BSON arrays.
+///
+class array {
+    template <bool B, class T = void>
+    using enable_if_t = typename std::enable_if<B, T>::type;
+
+    template <class T>
+    using decay_t = typename std::decay<T>::type;
+
+   public:
+    ///
+    /// Default constructor. Creates an empty array.
+    ///
+    array() = default;
+
+    ///
+    /// Converting constructor. Initializes each element of the array with the corresponding value
+    /// in std::forward<Args>(args).
+    ///
+    /// @param args values used to initialize each element of the array
+    ///
+    /// @note a bsoncxx::types::bson_value::value is direct-initialized from each argument.
+    ///
+    /// @see bsoncxx::types::bson_value::value
+    ///
+    template <typename... Args>
+    array(Args&&... args) {
+        this->_append(_core, std::forward<Args>(args)...);
+    }
+
+    ///
+    /// Move constructor. Constructs the array with the contents of other using move semantics.
+    ///
+    /// @param other another array to use as source to initialize the array with
+    ///
+    array(array&& other) noexcept = default;
+
+    ///
+    /// Replaces the contents with those of other using move semantics.
+    ///
+    /// @param other another array to use as source to initialize the array with
+    ///
+    array& operator=(array&& other) noexcept = default;
+
+    ///
+    /// Copy constructor. Constructs the array with a copy of the contents of other.
+    ///
+    /// @param other another array to use as source to initialize the array with
+    ///
+    array(const array& other) {
+        for (auto&& value : other._core.view_array())
+            _core.append(bson_value::value{value.get_value()});
+    }
+
+    ///
+    /// Replaces the contents with a copy of other. If *this and other are the same object, this
+    /// function has no effect.
+    ///
+    /// @param other another array to use as source to initialize the array with
+    ///
+    /// @return *this
+    ///
+    array& operator=(const array& other) {
```

Comment:
The existing core builder doesn't have a copy constructor or copy assignment operator.  It looks like you were able to work around that on L79-82 by manually appending the contents of `_core` to the new array builder instead of using a default.  But it makes me think about what it means to copy a builder, and I can't think of a use case where it would be desirable to copy this class.  What was the reason for adding the copy ctor and assignment operators?

Reply:
👍